### PR TITLE
Feature/delete local vol implementation

### PR DESCRIPTION
**Pull request by ChristianAtDell**
# Description
Driver-level implementation of DeleteLocalVolume on PowerStore.[ Corresponds to this PR](https://github.com/dell/csi-powerstore/pull/228), which added the stub. 

PR is set as draft for now.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| [665](https://github.com/dell/csm/issues/665) |

# Checklist:

- [X] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [X] I have verified that new and existing unit tests pass locally with my changes
- [X] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

Manual testing to verify that the backend volume is deleted. String of calls from driver logs are as follows:

Source side controller: 

```
{"level":"info","logger":"controllers.PersistentVolume","msg":"Deletion requested annotation not found","persistentvolume":"ccoff-csivol-bd07f02cb9","time":"2023-03-27T18:21:34.409949534Z"}
{"level":"info","logger":"controllers.PersistentVolume","msg":"Retention policy is set to Delete","persistentvolume":"ccoff-csivol-bd07f02cb9","time":"2023-03-27T18:21:34.412726422Z"}
{"level":"info","logger":"controllers.PersistentVolume","msg":"Adding deletion requested annotation to remote volume","persistentvolume":"ccoff-csivol-bd07f02cb9","time":"2023-03-27T18:21:34.412764601Z"}
{"level":"info","logger":"controllers.PersistentVolume","msg":"Adding sync delete annotation to remote","persistentvolume":"ccoff-csivol-bd07f02cb9","time":"2023-03-27T18:21:34.412775669Z"}
{"level":"info","logger":"controllers.PersistentVolume","msg":"Reconciling PV event","time":"2023-03-27T18:21:34.418976533Z"}
```

Target side replicator:

```
{"level":"info","logger":"controllers.PersistentVolume","msg":"Synchronized Deletion requested by annotation, deleting local backend volume","persistentvolume":"{\"Namespace\":\"\",\"Name\":\"ccoff-csivol-bd07f02cb9\"}","time":"2023-03-27T18:21:34.421301732Z"}
{"error":"Operation cannot be fulfilled on persistentvolumes \"ccoff-csivol-bd07f02cb9\": the object has been modified; please apply your changes to the latest version and try again","level":"error","logger":"controllers.PersistentVolume","msg":"Failed to add SynchronizedDeletionStatus complete annotation to the PV","persistentvolume":"{\"Namespace\":\"\",\"Name\":\"ccoff-csivol-bd07f02cb9\"}","time":"2023-03-27T18:21:35.28659685Z"}
{"controller":"persistentvolume","controllerGroup":"","controllerKind":"PersistentVolume","error":"Operation cannot be fulfilled on persistentvolumes \"ccoff-csivol-bd07f02cb9\": the object has been modified; please apply your changes to the latest version and try again","level":"error","msg":"Reconciler error","name":"ccoff-csivol-bd07f02cb9","namespace":"","persistentVolume":"{\"name\":\"ccoff-csivol-bd07f02cb9\"}","reconcileID":"\"122053ad-57fc-4edf-810a-abff1a316b9c\"","time":"2023-03-27T18:21:35.286662832Z"}
{"level":"info","logger":"controllers.PersistentVolume","msg":"Begin reconcile - PV Controller","persistentvolume":"{\"Namespace\":\"\",\"Name\":\"ccoff-csivol-bd07f02cb9\"}","time":"2023-03-27T18:21:35.286755626Z"}
{"level":"info","logger":"controllers.PersistentVolume","msg":"Adding replication-group, remote storage class annotation and label to the PersistentVolume","persistentvolume":"{\"Namespace\":\"\",\"Name\":\"ccoff-csivol-bd07f02cb9\"}","time":"2023-03-27T18:21:35.286835217Z"}
{"level":"debug","logger":"controllers.PersistentVolume","msg":"Checking if PV protection complete annotation is applied","persistentvolume":"{\"Namespace\":\"\",\"Name\":\"ccoff-csivol-bd07f02cb9\"}","time":"2023-03-27T18:21:35.286878691Z"}
{"level":"debug","logger":"controllers.PersistentVolume","msg":"Checking if synchronized deletion annotation is applied","persistentvolume":"{\"Namespace\":\"\",\"Name\":\"ccoff-csivol-bd07f02cb9\"}","time":"2023-03-27T18:21:35.286894258Z"}
{"level":"info","logger":"controllers.PersistentVolume","msg":"Synchronized Deletion requested by annotation, deleting local backend volume","persistentvolume":"{\"Namespace\":\"\",\"Name\":\"ccoff-csivol-bd07f02cb9\"}","time":"2023-03-27T18:21:35.286902148Z"}
```

Target side driver:

```
{"level":"info","msg":"/replication.v1.Replication/DeleteLocalVolume: REQ 0043: VolumeHandle=d76429e0-b1c7-497d-94f4-e78e7fd68498/PS721697ac108a/scsi, VolumeAttributes=map[arrayID:PS721697ac108a replication.storage.dell.com/ignoreNamespaces:false replication.storage.dell.com/isReplicationEnabled:true replication.storage.dell.com/remoteClusterID:ccoff-k8s replication.storage.dell.com/remotePVRetentionPolicy:Delete replication.storage.dell.com/remoteRGRetentionPolicy:Delete replication.storage.dell.com/remoteStorageClassName:pstore-rep-src replication.storage.dell.com/remoteSystem:WX-9023 replication.storage.dell.com/rpo:Five_Minutes replication.storage.dell.com/volumeGroupPrefix:ccoff-csi]","time":"2023-03-27T18:21:34.421746661Z"}
{"level":"info","msg":"Deleting local volume d76429e0-b1c7-497d-94f4-e78e7fd68498/PS721697ac108a/scsi per request from remote replication controller","time":"2023-03-27T18:21:34.421918765Z"}
{"level":"debug","msg":"REQUEST: GET /api/rest/volume/d76429e0-b1c7-497d-94f4-e78e7fd68498?select=%2A HTTP/1.1 Host: 10.230.24.67 Application-Type: CSI Driver for Dell EMC PowerStore/2.6.0+dirty Authorization: ******  ","time":"2023-03-27T18:21:34.422247109Z"}
{"level":"debug","msg":"acquire a lock","time":"2023-03-27T18:21:34.422347894Z"}
{"level":"debug","msg":"RESPONSE: HTTP/1.1 200 OK Content-Length: 1134 Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly X-Content-Type-Options: nosniff  {\"id\":\"d76429e0-b1c7-497d-94f4-e78e7fd68498\",\"name\":\"ccoff-csivol-bd07f02cb9\",\"description\":\"test-pvc-default\",\"type\":\"Primary\",\"wwn\":\"naa.68ccf098000fc284411fa4125e0d24bf\",\"nsid\":3176,\"nguid\":\"nguid.0fc284411fa4125e8ccf0968000d24bf\",\"appliance_id\":\"A1\",\"state\":\"Ready\",\"size\":5368709120,\"logical_used\":0,\"node_affinity\":\"System_Select_At_Attach\",\"creation_timestamp\":\"2023-03-27T17:32:25.047902+00:00\",\"protection_policy_id\":null,\"performance_policy_id\":\"default_medium\",\"is_replication_destination\":true,\"migration_session_id\":null,\"metro_replication_session_id\":null,\"is_host_access_available\":true,\"protection_data\":{\"family_id\": \"d76429e0-b1c7-497d-94f4-e78e7fd68498\", \"parent_id\": null, \"source_id\": null, \"creator_type\": \"User\", \"copy_signature\": null, \"source_timestamp\": \"2023-03-27T18:17:40.474399+00:00\", \"creator_type_l10n\": \"User\", \"is_app_consistent\": false, \"created_by_rule_id\": null, \"created_by_rule_name\": null, \"expiration_timestamp\": null},\"location_history\":null,\"app_type\":null,\"app_type_other\":null,\"type_l10n\":\"Primary\",\"state_l10n\":\"Ready\",\"node_affinity_l10n\":\"System Select At Attach\",\"app_type_l10n\":null}\n","time":"2023-03-27T18:21:34.501491619Z"}
{"level":"debug","msg":"release a lock","time":"2023-03-27T18:21:34.501990607Z"}
{"level":"debug","msg":"REQUEST: GET /api/rest/volume/d76429e0-b1c7-497d-94f4-e78e7fd68498?select=volume_group.volume_group_membership%28id%2Cname%2Cprotection_policy_id%29 HTTP/1.1 Host: 10.230.24.67 Application-Type: CSI Driver for Dell EMC PowerStore/2.6.0+dirty Authorization: ******  ","time":"2023-03-27T18:21:34.502078223Z"}
{"level":"debug","msg":"acquire a lock","time":"2023-03-27T18:21:34.502121033Z"}
{"level":"debug","msg":"RESPONSE: HTTP/1.1 200 OK Content-Length: 19 Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly X-Content-Type-Options: nosniff  {\"volume_group\":[]}\n","time":"2023-03-27T18:21:34.548834189Z"}
{"level":"debug","msg":"release a lock","time":"2023-03-27T18:21:34.548923433Z"}
{"level":"info","msg":"{[]}","time":"2023-03-27T18:21:34.549174853Z"}
{"level":"debug","msg":"REQUEST: DELETE /api/rest/volume/d76429e0-b1c7-497d-94f4-e78e7fd68498 HTTP/1.1 Host: 10.230.24.67 Application-Type: CSI Driver for Dell EMC PowerStore/2.6.0+dirty Authorization: ******  ","time":"2023-03-27T18:21:34.549288062Z"}
{"level":"debug","msg":"acquire a lock","time":"2023-03-27T18:21:34.549318168Z"}
{"level":"debug","msg":"RESPONSE: HTTP/1.1 204 No Content Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly X-Content-Type-Options: nosniff  \n","time":"2023-03-27T18:21:35.281151047Z"}
{"level":"debug","msg":"release a lock","time":"2023-03-27T18:21:35.281190205Z"}
{"level":"info","msg":"Local volume deleted successfully.","time":"2023-03-27T18:21:35.281197462Z"}
{"level":"info","msg":"/replication.v1.Replication/DeleteLocalVolume: REP 0043","time":"2023-03-27T18:21:35.281230758Z"}
{"level":"info","msg":"/replication.v1.Replication/DeleteLocalVolume: REQ 0044: VolumeHandle=d76429e0-b1c7-497d-94f4-e78e7fd68498/PS721697ac108a/scsi, VolumeAttributes=map[arrayID:PS721697ac108a replication.storage.dell.com/ignoreNamespaces:false replication.storage.dell.com/isReplicationEnabled:true replication.storage.dell.com/remoteClusterID:ccoff-k8s replication.storage.dell.com/remotePVRetentionPolicy:Delete replication.storage.dell.com/remoteRGRetentionPolicy:Delete replication.storage.dell.com/remoteStorageClassName:pstore-rep-src replication.storage.dell.com/remoteSystem:WX-9023 replication.storage.dell.com/rpo:Five_Minutes replication.storage.dell.com/volumeGroupPrefix:ccoff-csi]","time":"2023-03-27T18:21:35.287278331Z"}
{"level":"info","msg":"Deleting local volume d76429e0-b1c7-497d-94f4-e78e7fd68498/PS721697ac108a/scsi per request from remote replication controller","time":"2023-03-27T18:21:35.287311735Z"}
{"level":"debug","msg":"REQUEST: GET /api/rest/volume/d76429e0-b1c7-497d-94f4-e78e7fd68498?select=%2A HTTP/1.1 Host: 10.230.24.67 Application-Type: CSI Driver for Dell EMC PowerStore/2.6.0+dirty Authorization: ******  ","time":"2023-03-27T18:21:35.287440842Z"}
{"level":"debug","msg":"acquire a lock","time":"2023-03-27T18:21:35.287454159Z"}
{"level":"debug","msg":"RESPONSE: HTTP/1.1 404 Not Found Content-Length: 198 Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 X-Content-Type-Options: nosniff  {\"messages\":[{\"code\":\"0xE04040020009\",\"severity\":\"Error\",\"message_l10n\":\"Instance with id d76429e0-b1c7-497d-94f4-e78e7fd68498 was not found.\",\"arguments\":[\"d76429e0-b1c7-497d-94f4-e78e7fd68498\"]}]}\n","time":"2023-03-27T18:21:35.332258501Z"}
{"level":"debug","msg":"release a lock","time":"2023-03-27T18:21:35.33235969Z"}
{"level":"info","msg":"Volume does not exist. It may have already been deleted.","time":"2023-03-27T18:21:35.332375532Z"}
{"level":"info","msg":"/replication.v1.Replication/DeleteLocalVolume: REP 0044","time":"2023-03-27T18:21:35.332401186Z"}
```

Note: It looks like something in the replicator is failing to update the PV on the first pass, which causes DeleteLocalVolume to run twice. The second run is a NOP because the backend volume is already deleted, as per the last lines in the driver log. It should be addressed in the replicator.

### CODE REVIEW

File: pkg/controller/replication.go [hunk 1]
```diff
@@ -507,6 +509,64 @@ func (s *Service) DeleteStorageProtectionGroup(ctx context.Context,
...
+	splitHandle := strings.Split(req.VolumeHandle, `/`)
+	if len(splitHandle) != 3 {
+		return nil, status.Errorf(codes.InvalidArgument, "can't delete volume of improper handle format")
+	}
+	volumeID := splitHandle[0]
+	globalID := splitHandle[1]
+
+	arr, ok := s.Arrays()[globalID]
+	if !ok {
+		return nil, status.Errorf(codes.InvalidArgument, "can't find array with global id %s", globalID)
```
Comment (santhoshatdell) [inline pkg/controller/replication.go:527]: ID

File: pkg/controller/replication.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ package controller
 import (
```
Comment (santhoshatdell) [inline pkg/controller/replication.go:19]: Copyright year needs to be updated.

**Review by santhoshatdell — APPROVED** [review]

**Review by santhoshatdell — APPROVED** [review]

**Review by harshaatdell — APPROVED** [review]

File: pkg/controller/replication.go [hunk 3]
```diff
@@ -507,6 +509,66 @@ func (s *Service) DeleteStorageProtectionGroup(ctx context.Context,
 	return &csiext.DeleteStorageProtectionGroupResponse{}, nil
 }
 
+// DeleteLocalVolume deletes a volume on the local storage array upon request from a remote replication controller.
+func (s *Service) DeleteLocalVolume(ctx context.Context,
+	req *csiext.DeleteLocalVolumeRequest) (*csiext.DeleteLocalVolumeResponse, error) {
+	log.Info("Deleting local volume " + req.VolumeHandle + " per request from remote replication controller")
```
Comment (donatwork) [inline pkg/controller/replication.go:515]: Are any details/identity of the remote replication controller logged?

**Review by donatwork — APPROVED** [review]

File: pkg/controller/replication.go (hunk 3 above)
Comment (ChristianAtDell) [reply pkg/controller/replication.go:515]: Not here. We don't have access to information on the remote replication controller from the local replicator sidecar. Replicator can't communicate cross-cluster, only controller can.

The remote replication controller *does* log what it's doing, including setting the annotation that causes this delete, on the remote side.

File: pkg/controller/replication.go [hunk 4]
```diff
@@ -507,6 +509,66 @@ func (s *Service) DeleteStorageProtectionGroup(ctx context.Context,
...
+	if !ok {
+		return nil, status.Errorf(codes.InvalidArgument, "can't find array with global ID %s", globalID)
+	}
+
+	vol, err := arr.GetClient().GetVolume(ctx, volumeID)
+	if err != nil {
+		if apiError, ok := err.(gopowerstore.APIError); ok {
+			if apiError.NotFound() {
+				// volume doesn't exist, return success
+				log.Info(("Volume does not exist. It may have already been deleted."))
```
Comment (adarsh-dell) [inline pkg/controller/replication.go:535]: I didn't get the logic of using 2 braces '((' here?
Was it typo?

File: pkg/controller/controller_test.go [hunk 5]
```diff
@@ -3653,7 +3653,7 @@ var _ = Describe("CSIControllerService", func() {
 					Return(gopowerstore.Cluster{Name: validClusterName}, nil)
```
Comment (adarsh-dell) [inline pkg/controller/controller_test.go:3653]: Copyright year is:  © 2021-2022. Can you please change it to Copyright © 2021-2023?

**Comment by adarsh-dell** [comment]
@ChristianAtDell Would it be feasible to add UT for newly added function i.e. DeleteLocalVolume?
Thanks

**Comment by ChristianAtDell** [comment]
Unit tests updated, PR comments addressed. Should be good for review.

**Review by santhoshatdell — DISMISSED** [review]

File: pkg/controller/replication.go [hunk 6]
```diff
@@ -507,6 +509,66 @@ func (s *Service) DeleteStorageProtectionGroup(ctx context.Context,
...
+		log.Info("Cannot delete local volume " + volumeID + ", volume is part of a Volume Group and needs to be removed first.")
+		return nil, status.Errorf(codes.Internal, "Error: Unable to delete volume")
+	} else if vol.ProtectionPolicyID != "" {
+		log.Info("Cannot delete local volume " + volumeID + ", volume is under a protection policy that must be removed first.")
+		return nil, status.Errorf(codes.Internal, "Error: Unable to delete volume")
+	}
+
+	_, err = arr.GetClient().DeleteVolume(ctx, nil, volumeID)
+	if err != nil {
+		if apiErr, ok := err.(gopowerstore.APIError); ok && !apiErr.NotFound() {
```
Comment (santhoshatdell) [inline pkg/controller/replication.go:562]: Can you change the check to _!ok || !apiError.NotFound()_ ?

**Review by santhoshatdell — APPROVED** [review]

**Review by AkshaySainiDell — APPROVED** [review]

**Pull request merged**